File: cpp/0357.0_Count_Numbers_with_Unique_Digits.cpp
/*
执行用时：0 ms, 在所有 C++ 提交中击败了100.00% 的用户
内存消耗：5.7 MB, 在所有 C++ 提交中击败了99.61% 的用户
通过测试用例：9 / 9
*/
class Solution {
public:
    int countNumbersWithUniqueDigits(int n) {
        int ans = 1, init = 9, i = 1;
        while (i <= n) {
            ans += init;
            init *= (9 - i + 1);
            i += 1;
        }
        return ans;
    }
};


/*
执行用时：0 ms, 在所有 C++ 提交中击败了100.00% 的用户
内存消耗：5.7 MB, 在所有 C++ 提交中击败了96.99% 的用户
通过测试用例：9 / 9
*/
class Solution {
public:
    int countNumbersWithUniqueDigits(int n) {
       int ans = 0;
        switch(n) {
            case 8: ans += 9 * 9 * 8 * 7 * 6 * 5 * 4 * 3;
            case 7: ans += 9 * 9 * 8 * 7 * 6 * 5 * 4;
            case 6: ans += 9 * 9 * 8 * 7 * 6 * 5;
            case 5: ans += 9 * 9 * 8 * 7 * 6;
            case 4: ans += 9 * 9 * 8 * 7;
            case 3: ans += 9 * 9 * 8;
            case 2: ans += 9 * 9;
            case 1: ans += 9;
            case 0: ans += 1;
        }
        return ans;
    }
};

File: leetcode-cn/0357.0_Count_Numbers_with_Unique_Digits.py
'''
排列组合

执行用时：40 ms, 在所有 Python3 提交中击败了38.54% 的用户
内存消耗：14.9 MB, 在所有 Python3 提交中击败了52.56% 的用户
通过测试用例：9 / 9
'''
class Solution:
    def countNumbersWithUniqueDigits(self, n: int) -> int:
        c = [
            1,
            9,
            9 * 9,
            9 * 9 * 8,
            9 * 9 * 8 * 7,
            9 * 9 * 8 * 7 * 6,
            9 * 9 * 8 * 7 * 6 * 5,
            9 * 9 * 8 * 7 * 6 * 5 * 4,
            9 * 9 * 8 * 7 * 6 * 5 * 4 * 3,
        ]
        return sum(c[: n + 1])


'''
执行用时：40 ms, 在所有 Python3 提交中击败了38.54% 的用户
内存消耗：14.8 MB, 在所有 Python3 提交中击败了89.76% 的用户
通过测试用例：9 / 9
'''
class Solution:
    def countNumbersWithUniqueDigits(self, n: int) -> int:
        ans, init = 1, 9
        i = 1 
        while i <= n:
            ans += init 
            init *= (9 - i + 1)
            i += 1
        return ans 

